src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,360 @@
+{
+    if (aValue.data() != nullptr)
+    {
+        for (auto & item : aValue)
+        {
+            CleanUpIDs(item.tariffComponentIDs);
+        }
+
+        Platform::MemoryFree(aValue.data());
+        aValue = DataModel::List<Structs::MeteredQuantityStruct::Type>();
+    }
+}
+
+Instance::~Instance()
+{
+    if (!mMeteredQuantity.IsNull())
+    {
+        CleanupMeteredQuantityData(mMeteredQuantity.Value());
+    }
+    Shutdown();
+}
+
+CHIP_ERROR Instance::Init()
+{
+    mMeteredQuantity.SetNull();
+    mMeteredQuantityTimestamp.SetNull();
+    mMeasurementType.SetNull();
+    mMaximumMeteredQuantities.SetNull();
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+static CHIP_ERROR CopyMeteredQuantityEntry(const Structs::MeteredQuantityStruct::Type & src,
+                                           Structs::MeteredQuantityStruct::Type & dest)
+{
+    dest.quantity = src.quantity;
+
+    if (!SpanCopier<uint32_t>::Copy(src.tariffComponentIDs, dest.tariffComponentIDs, kMaxTariffComponentIDsPerMeteredQuantityEntry))
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetMeteredQuantity(const DataModel::Nullable<DataModel::List<Structs::MeteredQuantityStruct::Type>> & newValue)
+{
+    assertChipStackLockedByCurrentThread();
+
+    if (NullableListEqual(newValue, mMeteredQuantity))
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (newValue.IsNull())
+    {
+        mMeteredQuantity.SetNull();
+    }
+    else
+    {
+        if (!mMeteredQuantity.IsNull())
+        {
+            CleanupMeteredQuantityData(mMeteredQuantity.Value());
+        }
+
+        const size_t len = newValue.Value().size();
+
+        if (len == 0)
+        {
+            mMeteredQuantity = MakeNullable(DataModel::List<Structs::MeteredQuantityStruct::Type>());
+        }
+        else if (len <= kMaxMeteredQuantityEntries)
+        {
+            Platform::ScopedMemoryBuffer<Structs::MeteredQuantityStruct::Type> buffer;
+
+            if (!buffer.Calloc(len))
+            {
+                return CHIP_ERROR_NO_MEMORY;
+            }
+
+            for (size_t idx = 0; idx < len; idx++)
+            {
+                // Deep copy each MeteredQuantityStruct in the newValue list
+                ReturnLogErrorOnFailure(CopyMeteredQuantityEntry(newValue.Value()[idx], buffer[idx]));
+            }
+
+            mMeteredQuantity = MakeNullable(DataModel::List<Structs::MeteredQuantityStruct::Type>(buffer.Get(), len));
+            (void) buffer.Release();
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_LIST_LENGTH;
+        }
+
+        SetMaximumMeteredQuantities(MakeNullable(static_cast<uint16_t>(len)));
```

Comment:
This doesn't make sense to me.  MaximumMeteredQuantities is the maximum size of MeteredQuantity, not the current size, in the spec....  I would not expect it to change just because MeteredQuantity changes.